src/values/acceleration-structure.cxx

```diff
@@ -0,0 +1,1484 @@
+            const unsigned strideSBT = 0,
+            const unsigned missIndex = 0) {
+        
+        assert(rayOrigin.size() == 3 && rayDirection.size() == 3);
+        glm::vec4 convertedRayOrigin = glm::make_vec4(rayOrigin.data());
+        convertedRayOrigin.w = 1.0f;
+        glm::vec4 convertedRayDirection = glm::make_vec4(rayDirection.data());
+        convertedRayDirection.w = 0.0f;
+
+        this->rayFlags = rayFlags;
+        this->cullMask = cullMask;
+        this->rayOrigin = convertedRayOrigin;
+        this->rayDirection = convertedRayDirection;
+        this->rayTMin = rayTMin;
+        this->rayTMax = rayTMax;
+
+        this->useSBT = useSBT;
+        this->offsetSBT = offsetSBT;
+        this->strideSBT = strideSBT;
+        this->missIndex = missIndex;
+
+        setFlags();
+        resetTrace();
+    }
+
+    void initTrace(const unsigned& rayFlags,
+            const unsigned& cullMask,
+            const glm::vec4& rayOrigin,
+            const glm::vec4& rayDirection,
+            const float& rayTMin,
+            const float& rayTMax,
+            const bool& useSBT,
+            const unsigned offsetSBT = 0,
+            const unsigned strideSBT = 0,
+            const unsigned missIndex = 0) {
+                
+        this->rayFlags = rayFlags;
+        this->cullMask = cullMask;
+        this->rayOrigin = rayOrigin;
+        this->rayDirection = rayDirection;
+        this->rayTMin = rayTMin;
+        this->rayTMax = rayTMax;
+
+        this->useSBT = useSBT;
+        this->offsetSBT = offsetSBT;
+        this->strideSBT = strideSBT;
+        this->missIndex = missIndex;
+
+        setFlags();
+        resetTrace();
+    }
+
+private:
+    bool didPopNodePreviously = true;
+
+public:
+    // TODO: step to the next primitive
+    // TODO: change instance node pop and push logic
+    bool stepTrace(bool& didIntersectGeometry) {
+        if (!activeTrace) {
+            std::cout << "Out of paths to trace ray for acceleration structure id = (" << id << ")." << std::endl
+                      << "Please (re-)initialize (or reset) the trace." << std::endl;
+            return false;
+        }
+
+        bool foundPrimitive = false;
+
+        while (!foundPrimitive && (nodesToEval.size() > 0)) {
+
+            NodeRef currNodeRef = nodesToEval.top();
+            Node* currNode = currNodeRef->get();
+            nodesToEval.pop();
+
+            std::cout << "Trace ray: node type: " << static_cast<unsigned>(currNode->type()) << std::endl;
+
+            switch(currNode->type()) {
+            default: {
+                std::stringstream err;
+                err << "Found unknown node type (" << static_cast<int>(currNode->type())
+                    << ") in \"traceRay()\" method of class "
+                        "\"AccelerationStructure\" in \"acceleration-structure.cxx\"";
+                throw std::runtime_error(err.str());
+            }
+            case NodeType::Box: {
+                // TODO
```

Comment:
What remains to be done here? There is a TODO but no explanation and significant amount of code beneath it.